apps/lib/app_provider.c

```diff
@@ -69,6 +89,8 @@ int opt_provider(int opt)
         return opt_provider_load(opt_arg());
     case OPT_PROV_PROVIDER_PATH:
         return opt_provider_path(opt_arg());
+    case OPT_PROV_PROVIDER_CONFIG_FILE:
+        return opt_libctx_config_path(opt_arg());
```

Comment:
I'm not too sure about the semantics of this. Effectively this introduces the side effect of creating a non-default libctx. This is not at all obvious and means that `-provider default -provider_config conf.cnf` means something different to `-provider_config conf.cnf -provider default`. This seems quite non-intuitive. It also means the option will be ignored for any app that we don't update to being "libctx" aware.

A possible alternative is that these options never create a new libctx themselves. They only use whatever the value of `libctx` is. Applications that are libctx aware could call a `setup_libctx` function which creates a non-default value for libctx, and loads the "null" provider into the default ctx.

Reply:
Hmm I just tried doing this in #11948..
So having the app always have the libctx is also not good (at least initially as it propogates to every other instance of the app being called). if we have just one case where the libctx is not valid (e.g: SM2) then another test will break.. Maybe a -use_libctx option is better?

Reply:
Yes, that could work.

Reply:
ok done.